R/ui.R

```diff
@@ -0,0 +1,132 @@
+    callback(ui_url)
+  }
+
+  return(invisible())
+
+}
+
+#' Mount OpenAPI Specification to a plumber router
+#' @noRd
+mountOpenAPI <- function(pr, api_url) {
+
+  spec <- pr$swaggerFile()
+
+  # Create a function that's hardcoded to return the OpenAPI specification -- regardless of env.
+  openapi_fun <- function(req) {
+    # use the HTTP_REFERER so RSC can find the swagger location to ask
+    ## (can't directly ask for 127.0.0.1)
+    if (is.null(getOption("plumber.apiURL")) &&
+        is.null(getOption("plumber.apiHost"))) {
+      if (is.null(req$HTTP_REFERER)) {
+        # Prevent leaking host and port if option is not set
+        api_url <- character(1)
+      }
+      else {
+        # Use HTTP_REFERER as fallback
+        api_url <- req$HTTP_REFERER
+        api_url <- sub("index\\.html$", "", api_url)
+        api_url <- sub(paste0("__(", paste0(names(.globals$interfaces), collapse = "|"),")__/$"), "", api_url)
+      }
+    }
+
+    utils::modifyList(list(servers = list(list(url = api_url))), spec)
+
+  }
+  # http://spec.openapis.org/oas/v3.0.3#document-structure
+  # "It is RECOMMENDED that the root OpenAPI document be named: openapi.json"
+  pr$handle("GET", "/openapi.json", openapi_fun, serializer = serializer_unboxed_json())
+
+  return(invisible())
+
+}
+
+#' Mount Interface UI
+#' @noRd
+mountInterfaces <- function(interface) {
+  mountInterface <- function(pr, api_url, ...) {
+    if (!requireNamespace(interface$package, quietly = TRUE)) {
+      stop(interface$package, " must be installed for the ", interface$name," UI to be displayed")
+    }
+
+    interfacePath <- paste0("/__", tolower(interface$name), "__/")
+    interfaceUrl <- paste0(api_url, interfacePath)
+
+    html_content <- interface$index(...)
+    interface_index <- function() {
+      html_content
+    }
+    for (path in paste0(interfacePath, c("index.html", ""))) {
+      pr$handle(
+        "GET", path, interface_index,
+        serializer = serializer_html()
+      )
+    }
+    pr$mount(interfacePath, PlumberStatic$new(interface$static()))
+    return(interfaceUrl)
+  }
+  .globals$interfaces[[interface$name]] <- mountInterface
+}
+
+swaggerInterface <- list(
+  package = "swagger",
+  name = "swagger",
+  index = function() {
+    swagger::swagger_spec(
+      api_path = 'window.location.origin + window.location.pathname.replace(/\\(__swagger__\\\\/|__swagger__\\\\/index.html\\)$/, "") + "openapi.json"',
+      version = "3"
+    )
+  },
+  static = swagger::swagger_path
+)
+
+# redocInterface <- list(
+#   package = "redoc",
+#   name = "redoc",
+#   index = function(redoc_options = structure(list(), names = character())) {
+#     redoc::redoc_spec(
+#       spec_url = "\' + window.location.origin + window.location.pathname.replace(/\\(__redoc__\\\\/|__redoc__\\\\/index.html\\)$/, '') + 'openapi.json' + \'",
+#       redoc_options = redoc_options
+#     )
+#   },
+#   static = redoc::redoc_path
+# )
+
+mountInterfaces(swaggerInterface)
```

Comment:
This is where I'm not sure how to handle it. `mountInterfaces` is from plumber. Where should this particular part so that interface is available when needed.

Currently interfaces are added to globals on package load but they could also be added only when user use the $run method with a ui.

Reply:
There could be a public method in each package. Ex: `swagger::plumber_mount_interface()` 

This method could also be run in `.onLoad()`?

```r

plumber_mount_interface <- function() {
  plumber::mountInterface(....)
}

.onLoad <- function(...) {
  if (require("plumber", quietly = TRUE)) {
    plumber_mount_interface()
  }

}
```


> Currently interfaces are added to globals on package load

As long as they say `library(MYINTERFACEPKG)` in the plumb file, then it'll get picked up.

This setup will also help when APIs are deployed so packrat finds pkg `MYINTERFACEPKG`. If it's only a variable, packrat won't bundle it when deployed.

Reply:
This is going in redoc and swagger package. It should be non blocking here. I'll pr them some time soon.

Reply:
Pr done in swagger, redoc updated, works from 2.0.0.30